src/python/pants/backend/python/goals/pytest_runner_integration_test.py

```diff
@@ -3,421 +3,451 @@
+    file = digest_contents[0]
+    assert file.path.startswith("dist/test-results")
+    assert b"pants_test.test_good" in file.content
+
+
+def test_coverage(rule_runner: RuleRunner) -> None:
+    tgt = create_test_target(rule_runner, [GOOD_SOURCE])
+    result = run_pytest(rule_runner, tgt, use_coverage=True)
+    assert result.exit_code == 0
+    assert f"{PACKAGE}/test_good.py ." in result.stdout
+    assert result.coverage_data is not None
+
+
+def test_conftest_handling(rule_runner: RuleRunner) -> None:
+    """Tests that we a) inject a dependency on conftest.py and b) skip running directly on
+    conftest.py."""
+    tgt = create_test_target(rule_runner, [GOOD_SOURCE])
+
+    rule_runner.create_file(
+        f"{SOURCE_ROOT}/conftest.py", "def pytest_runtest_setup(item):\n  print('In conftest!')\n"
+    )
+    rule_runner.add_to_build_file(SOURCE_ROOT, "python_tests()")
+    conftest_tgt = rule_runner.get_target(Address(SOURCE_ROOT, relative_file_path="conftest.py"))
+    assert isinstance(conftest_tgt, PythonTests)
+
+    result = run_pytest(rule_runner, tgt, passthrough_args="-s")
+    assert result.exit_code == 0
+    assert f"{PACKAGE}/test_good.py In conftest!\n." in result.stdout
+
+    result = run_pytest(rule_runner, conftest_tgt)
+    assert result.exit_code is None
+
+
+def test_execution_slot_variable(rule_runner: RuleRunner) -> None:
+    source = FileContent(
+        path=f"{PACKAGE}/test_concurrency_slot.py",
+        content=dedent(
+            """\
+            import os
+
+            def test_fail_printing_slot_env_var():
+                slot = os.getenv("SLOT")
+                print(f"Value of slot is {slot}")
+                # Deliberately fail the test so the SLOT output gets printed to stdout
+                assert 1 == 2
+            """
+        ).encode(),
+    )
+    tgt = create_test_target(rule_runner, [source])
+    result = run_pytest(rule_runner, tgt, execution_slot_var="SLOT")
+    assert result.exit_code == 1
+    assert re.search(r"Value of slot is \d+", result.stdout)
+
+
+def test_extra_env_vars(rule_runner: RuleRunner) -> None:
+    source = FileContent(
+        path=f"{PACKAGE}/test_extra_env_vars.py",
+        content=dedent(
+            """\
+            import os
+
+            def test_args():
+                assert os.getenv("SOME_VAR") == "some_value"
+                assert os.getenv("OTHER_VAR") == "other_value"
+            """
+        ).encode(),
+    )
+    tgt = create_test_target(rule_runner, [source])
+    mock_env = PantsEnvironment({"OTHER_VAR": "other_value"})
+    extra_env_vars = '["SOME_VAR=some_value", "OTHER_VAR"]'
+    result = run_pytest(rule_runner, tgt, extra_env_vars=extra_env_vars, pants_environment=mock_env)
+    assert result.exit_code == 0
+
+
+@pytest.mark.skip(
+    reason=(
+        "TODO: Figure out why the rule isn't re-evaluated each time, even if `--no-pantsd` is set."
+    )
+)
```

Comment:
I couldn't figure this one out. I tested that this same set up works in real code, so I think it's a quirk with `RuleRunner`.

Reply:
Huh...